space_packet_parser/packets.py

```diff
@@ -29,13 +30,67 @@ class SequenceFlags(IntEnum):
 
 
 class RawPacketData(bytes):
-    """A class to represent raw packet data as bytes but whose length is represented by bit length.
+    """A class containing the raw binary packet data.
 
     This class is a subclass of bytes and is used to represent the raw packet data
-    in a more readable way. It is used to store the raw packet data in the Packet
+    in a more readable way. It is used to store the binary data in the Packet
     class and used to keep track of the current parsing position (accessible through the `pos` attribute).
+    """
+    pos = 0  # Current position in bits
+
+    def read_as_bytes(self, nbits: int) -> bytes:
+        """Read a number of bits from the packet data as bytes. Reads minimum number of complete bytes required to
+        capture `nbits`. Moves `pos` cursor `nbits` forward, even if `nbits` is not an integer number of bytes.
+
+        Parameters
+        ----------
+        nbits : int
+            Number of bits to read
+
+        Returns
+        -------
+        : bytes
+            Raw bytes from the packet data
+        """
+        if self.pos + nbits > len(self) * 8:
+            raise ValueError("Tried to read beyond the end of the packet data. "
+                             f"Tried to read {nbits} bits from position {self.pos} "
+                             f"in a packet of length {len(self) * 8} bits.")
+        if self.pos % 8 == 0 and nbits % 8 == 0:
+            # If the read is byte-aligned, we can just return the bytes directly
+            data = self[self.pos//8:self.pos//8 + (nbits+7) // 8]
+            self.pos += nbits
+            return data
+        # We are non-byte aligned, so we need to extract the bits and convert to bytes
+        bytes_as_int = _extract_bits(self, self.pos, nbits)
+        self.pos += nbits
+        return int.to_bytes(bytes_as_int, (nbits + 7) // 8, "big")
+
+    def read_as_int(self, nbits: int) -> int:
+        """Read a number of bits from the packet data as an integer.
+
+        Parameters
+        ----------
+        nbits : int
+            Number of bits to read
 
-    There are also a few convenience methods to extract the CCSDS header fields from the packet data.
+        Returns
+        -------
+        : int
+            Integer representation of the bits read from the packet
+        """
+        if self.pos + nbits > len(self) * 8:
+            raise ValueError("Tried to read beyond the end of the packet data. "
+                             f"Tried to read {nbits} bits from position {self.pos} "
+                             f"in a packet of length {len(self) * 8} bits.")
+        int_data = _extract_bits(self, self.pos, nbits)
+        self.pos += nbits
+        return int_data
+
+class CCSDSPacketBytes(bytes):
```

Comment:
Intuitively, I would expect `CCSDSPacketBytes` to subclass `RawPacketData` and to be called `CCSDSRawPacketData` to make it clear that they are related. Perhaps I misunderstand but both `RawPacketData` and `CCSDSPacketBytes` are just wrappers around `bytes`, but every method for `RawPacketData` should also be applicable to `CCSDSPacketBytes`, right?

Maybe more simply, `RawPacketData(bytes)` -> `PacketData(bytes)` (data meaning bytes), and `CCSDSPacketBytes(bytes)` -> `CCSDSPacketData(PacketData)`?

`CCSDSPacketData` would be strictly Liskov in that every `CCSDSPacketData` is a valid `PacketData` object because we're not adding limitations, only functionality (for parsing header fields as properties).

Reply:
Said in the other comment, but I don't think that the `CCSDSPacketBytes` should be "readable". It is just a stream of bytes, and there are some extra methods that do some nice things on the header bits for ease of access. The reader stuff within `RawPacketData` is really only useful for XTCE-parsing things. This is external to that and someone can just iterate over the raw bytes to count packets etc.. without any reading (like we are doing in some of the cli stuff).

Suggestions for improving this are welcome, but I don't think we should require someone to bring a `RawPacketData` subclass to the definition parsing, we should help them out and add the readers ourselves since that is really only meant for us.